src/linker/Linker/Driver.cs

```diff
@@ -674,11 +674,20 @@ public bool Run (ILogger customLogger = null)
 
 			try {
 				p.Process (context);
+			} catch (Exception ex) {
+				Console.Error.WriteLine ("IL Linker exited with error:");
+				if (ex is LinkerFatalErrorException le) {
+					context.LogMessage (le.MessageContainer);
+					if (le.InnerException != null)
+						Console.Error.WriteLine (le.InnerException);
+				} else {
+					Console.Error.WriteLine (ex);
```

Comment:
Why don't we write the exception to the error log. As is this would not report it as MSBuild error - right? That seems wrong. (Not counting that it breaks redirecting logs via interface).

In general I don't think linker should write directly to Console ever. But we can fix that in a separate change.
Since linker supports the `ILogger` all logging should go through that.

Reply:
If we use the logging interface for both `LinkerFatalErrorException` and `InternalErrorException`, what's the difference between the two? I thought the point of the internal exception was that we don't create error codes for it because they should never be seen by users (and if they are, it's a bug). We just want to see a stacktrace if the linker crashes.

Anything logged to stderr will currently go to the MSBuild output with MSBuild `MessageImportance.High`.

Reply:
I was thinking about `InternalErrorException` as a kind of dead end in the sense that we'd only see it if something went really wrong and we needed to kill the linker (there's a bug; we reached some code that wasn't supposed to be reached). Giving it a second thought, I think we might simply rely on `Debug.Assert` and existing exceptions for this, and add a new error code asking the users to file an issue with the stack trace.

Reply:
Using `InternalErrorException` or existing exceptions+`Debug.Assert` makes sense to me for the dead end behavior that you described. I don't think we need a new code though - I think of the error codes as user-facing errors - so I wouldn't force the internal errors to go through the same path that reports errors about the inputs. That might just be me - it sounds like @vitek-karas might have a different opinion. As long as the intended use of the exceptions is clear, I'm fine with either way.

Reply:
I don't know what the msbuild experience will be if we don't report an error code. As long as there's a clear message saying "Internal error, please report it to us" and then the details in the "Error Window" in VS, I'm fine with whichever solution we pick.

Reply:
If the linker fails during publish, this is what it currently looks like:
<img width="673" alt="Screen Shot 2020-05-06 at 12 49 12 PM" src="https://user-images.githubusercontent.com/787361/81222234-d1730b00-8f98-11ea-81cf-59043dc2716a.png">
The log mentioned in that popup doesn't have anything useful, but the output has the text written by the linker ("Expecting a parameter"). I haven't checked, but I think the new errors will show up similarly - they'll be errors in MSBuild, but won't make it to the VS error window if run during publish. I'll check this behavior later.

Reply:
Hmm - just wondering - what is the warning experience in that case? Does it even show up?

Reply:
I checked - yeah, it's the same for warnings and errors:
<img width="535" alt="Screen Shot 2020-05-06 at 3 17 07 PM" src="https://user-images.githubusercontent.com/787361/81234107-f83b3c80-8fac-11ea-9ac9-974f274b9b86.png">
Errors will additionally produce that popup.

Reply:
We should plan as if we were running during build. If I set it up to run during build, I see:
- if the task fails (should never happen)
<img width="472" alt="Screen Shot 2020-05-06 at 3 24 04 PM" src="https://user-images.githubusercontent.com/787361/81234689-1d7c7a80-8fae-11ea-96b1-b0cc235de752.png">

- if the linker fails (should never happen)
<img width="470" alt="Screen Shot 2020-05-06 at 3 26 20 PM" src="https://user-images.githubusercontent.com/787361/81234698-1fded480-8fae-11ea-84eb-e2c53720599a.png">
<img width="493" alt="Screen Shot 2020-05-06 at 3 26 30 PM" src="https://user-images.githubusercontent.com/787361/81234705-21a89800-8fae-11ea-813e-723d779c8b32.png">